File: src/main/java/ru/falseteam/neural2048/players/NeuralNetworkPlayer.java
package ru.falseteam.neural2048.players;

import org.jetbrains.annotations.NotNull;
import ru.falseteam.neural2048.logic.Directions;
import ru.falseteam.neural2048.logic.GameLogic;
import ru.falseteam.neural2048.logic.GameState;
import ru.falseteam.neural2048.nn.NeuralNetwork;

import java.util.*;

/**
 * Игрок играющий с помощью нейронной сети.
 * Сеть должна иметь 16 входных нейронов и 4 выходных нейрона.
 *
 * @author Vladislav Sumin
 * @version 2.0
 */
public class NeuralNetworkPlayer implements Player {
    private NeuralNetwork nn;

    private final Pair pairUp = new Pair(Directions.UP);
    private final Pair pairDown = new Pair(Directions.DOWN);
    private final Pair pairLeft = new Pair(Directions.LEFT);
    private final Pair pairRight = new Pair(Directions.RIGHT);
    private final Pair[] pairs = {pairUp, pairDown, pairLeft, pairRight};


    public NeuralNetworkPlayer(NeuralNetwork nn) {
        this.nn = nn;
    }

    @Override
    public void playOneGame(GameLogic gameLogic) {
        while (gameLogic.state == GameState.GAME) {
            //Установка входных значений нейронов
            for (int i = 0; i < 16; ++i) {
                nn.putSignalToNeuron(i, gameLogic.theGrid[i / 4][i % 4]);
            }

            //Активация
            nn.activate();

            //Снятие значений с выходных нейронов
            pairUp.value = nn.getAfterActivationSignal(nn.getNeuronsCount() - 1);
            pairDown.value = nn.getAfterActivationSignal(nn.getNeuronsCount() - 2);
            pairLeft.value = nn.getAfterActivationSignal(nn.getNeuronsCount() - 3);
            pairRight.value = nn.getAfterActivationSignal(nn.getNeuronsCount() - 4);

            //Обработка хода
            Arrays.sort(pairs);
            for (Pair move : pairs) {
                if (gameLogic.move(move.direction)) break;
            }
        }
    }

    public void setNeuralNetwork(NeuralNetwork nn) {
        this.nn = nn;
    }

    /**
     * Вспомогательный класс для корректной обработки приоретета ходов.
     */
    private static class Pair implements Comparable<Pair> {
        final Directions direction;
        Double value;

        Pair(Directions direction) {
            this.direction = direction;
        }

        @Override
        public int compareTo(@NotNull Pair o) {
            //Сортировка по убыванию
            return Double.compare(o.value, value);
        }
    }
}

File: src/main/java/ru/falseteam/neural2048/players/NeuralNetworkPlayer.kt
package ru.falseteam.neural2048.players

import ru.falseteam.neural2048.logic.Directions
import ru.falseteam.neural2048.logic.GameLogic
import ru.falseteam.neural2048.logic.GameState
import ru.falseteam.neural2048.nn.NeuralNetwork

import java.util.*

/**
 * Игрок, играющий с помощью нейронной сети.
 * Сеть должна иметь 16 входных и 4 выходных нейрона.
 *
 * @author Vladislav Sumin
 * @version 2.0
 */
class NeuralNetworkPlayer(private var nn: NeuralNetwork?, gameLogic: GameLogic) : Player(gameLogic) {

    private val pairUp = Pair(Directions.UP)
    private val pairDown = Pair(Directions.DOWN)
    private val pairLeft = Pair(Directions.LEFT)
    private val pairRight = Pair(Directions.RIGHT)
    private val pairs = arrayOf(pairUp, pairDown, pairLeft, pairRight)

    override fun playOneGame() {
        val nn = nn ?: throw IllegalStateException()
        while (gameLogic.state == GameState.GAME) {
            val maxTile: Double = gameLogic.maxTileExp.toDouble()
            // Установка входных значений нейронов
            for (i in 0..15) {
                nn.putSignalToNeuron(i, gameLogic.theGrid[i / 4][i % 4].toDouble() / maxTile)
            }

            // Активация
            nn.activate()

            // Снятие значений с выходных нейронов
            pairUp.value = nn.getAfterActivationSignal(nn.neuronsCount - 1)
            pairDown.value = nn.getAfterActivationSignal(nn.neuronsCount - 2)
            pairLeft.value = nn.getAfterActivationSignal(nn.neuronsCount - 3)
            pairRight.value = nn.getAfterActivationSignal(nn.neuronsCount - 4)

            // Обработка хода
            Arrays.sort(pairs)
            @Suppress("LoopToCallChain")
            for (move in pairs) {
                if (gameLogic.move(move.direction)) break
            }
        }
    }

    fun setNeuralNetwork(nn: NeuralNetwork) {
        this.nn = nn
    }

    /**
     * Вспомогательный класс для корректной обработки приоретета ходов.
     */
    private class Pair(val direction: Directions) : Comparable<Pair> {
        var value: Double? = null

        override fun compareTo(other: Pair): Int {
            //Сортировка по убыванию
            return java.lang.Double.compare(other.value!!, value!!)
        }
    }
}
